module/state_synchronization/requester/oneshot_execution_data_requester.go

```diff
@@ -0,0 +1,182 @@
+// Returns the execution data entity and any error encountered.
+//
+// The function logs each retry attempt and emits metrics for retries and fetch durations.
+// The block height is used only for logging and metric purposes.
+//
+// Expected errors:
+// - context.Canceled: if the provided context was canceled before completion
+// All other errors are unexpected exceptions and may indicate invalid execution data was received.
+func (r *OneshotExecutionDataRequester) RequestExecutionData(
+	ctx context.Context,
+) (*execution_data.BlockExecutionDataEntity, error) {
+	backoff := retry.NewExponential(r.config.RetryDelay)
+	backoff = retry.WithCappedDuration(r.config.MaxRetryDelay, backoff)
+	backoff = retry.WithJitterPercent(15, backoff)
+
+	// bitswap always waits for either all data to be received or a timeout, even if it encountered an error.
+	// use an incremental backoff for the timeout so we do faster initial retries, then allow for more
+	// time in case data is large or there is network congestion.
+	timeout := retry.NewExponential(r.config.FetchTimeout)
+	timeout = retry.WithCappedDuration(r.config.MaxFetchTimeout, timeout)
+
+	attempt := 0
+	blockID := r.executionResult.BlockID
+	blockHeight := r.blockHeader.Height
+	executionDataID := r.executionResult.ExecutionDataID
+
+	var execData *execution_data.BlockExecutionDataEntity
+	err := retry.Do(ctx, backoff, func(context.Context) error {
+		if attempt > 0 {
+			r.log.Debug().
+				Str("block_id", blockID.String()).
+				Str("execution_data_id", executionDataID.String()).
+				Uint64("height", blockHeight).
+				Uint64("attempt", uint64(attempt)).
+				Msgf("retrying download")
+
+			r.metrics.FetchRetried()
+		}
+		attempt++
+
+		// download execution data for the block
+		fetchTimeout, _ := timeout.Next()
+		var err error
+		execData, err = r.processFetchRequest(ctx, blockHeight, fetchTimeout)
+		if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) {
+			return retry.RetryableError(err)
+		}
+
+		if execution_data.IsMalformedDataError(err) || execution_data.IsBlobSizeLimitExceededError(err) {
+			// these errors indicate the execution data was received successfully and its hash matched
+			// the value in the ExecutionResult, however, the data was malformed or invalid. this means that
+			// an execution node produced an invalid execution data blob, and verification nodes approved it
+			r.log.Error().Err(err).Str(
+				"execution_data_id",
+				executionDataID.String(),
+			).Msg("received invalid execution data from network (potential slashing evidence?)")
+		}
+
+		return err
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return execData, nil
+}
+
+// processFetchRequest performs the actual fetch of execution data for the given block within the provided timeout.
+//
+// It wraps the fetch with metrics tracking and contextual logging. The execution data is retrieved
+// from the execution data cache based on the block ID.
+//
```

Comment:
this not accurate and I don't think we need the logging and metrics behavior in the public API docs. let's remove it
```suggestion
```